Task/document query parsing

jack-gitdev opened the pull request:
Add tests for verifying expected order of operations in absence of parenthesis. Add convenience methods for interacting with JexlNode objects.

jack-gitdev committed:
wip

jack-gitdev committed:
Added JexlASTHelper tests to verify expected order of operations in absence of parenthesis. Added convenience methods to JexlASTHelper for interacting with JexlNode objects.

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java:
```diff
@@ -1237,6 +1256,44 @@ public static boolean isWithinAnd(JexlNode node) {
     }
     
+    public static <T extends JexlNode> boolean isWithinNodeType(JexlNode node, Class<T> typeKey) {
+        while (null != node && null != node.jjtGetParent()) {
```
This returns after a single iteration, so there's no need for a while loop.  You didn't create them, but this is also true for isWithinAnd, and isWithinNot.

jack-gitdev replied:
Sounds good, I will change this is.

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java:
Javadoc to describe what your methods are doing would be appreciated.

jack-gitdev replied:
Will add Javadocs.

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java:
```diff
@@ -1237,6 +1256,44 @@ public static boolean isWithinAnd(JexlNode node) {
         return false;
     }
     
+    public static <T extends JexlNode> boolean isWithinNodeType(JexlNode node, Class<T> typeKey) {
```
I know there are already a few methods named 'isWithin...' but to me all of these methods would be better named as 'isDescendantOf...'

jack-gitdev replied:
Agreed, will change them.

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java:
```diff
@@ -1237,6 +1256,44 @@ public static boolean isWithinAnd(JexlNode node) {
+            for (int i = 0; i < numChildren; i++) {
+                JexlNode child = node.jjtGetChild(i);
+                
+                boolean foundType = false;
+                
+                if (typeKey.isInstance(child)) {
+                    return true;
+                } else {
+                    foundType = childrenContainNodeType(child, typeKey);
+                }
+                
+                if (foundType) {
+                    return true;
+                }
```
could be simplified

```suggestion
                if (typeKey.isInstance(child) || childrenContainNodeType(child, typeKey)) {
                    return true;
                }
```

jack-gitdev replied:
Thank you for this, I could not figure out how to simplify it but I know it didn't look great.

jwomeara reviewed (changes requested):
A few comments...

So we're adding these new methods, but they aren't being used anywhere outside of tests.  Is this something we foresee us needing as part of a future task?  Why do we need these methods?

jack-gitdev commented:
I will make the recommended changes. I added the convenience methods for my own use in the new tests. I didn't see any methods for checking against certain node types so I added these  generalized methods that can be used to check against any of the JexlNode based classes.

jack-gitdev committed:
Added JexlASTHelper tests to verify expected order of operations in absence of parenthesis for lucene queries. Added Javadocs and comments. Updated method names.

jwomeara reviewed (approved):
I think you covered all of my concerns.  LGTM

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java:
```diff
@@ -796,6 +820,43 @@ private static void getERNodes(JexlNode node, List<ASTERNode> erNodes) {
+    /**
+     * Iterate through provided node and its children, then return a list of nodes matching the provided class. The provided class must extend JexlNode.
+     * 
+     * @see org.apache.commons.jexl2.parser.JexlNode;
```
here

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java:
```diff
@@ -796,6 +820,43 @@ private static void getERNodes(JexlNode node, List<ASTERNode> erNodes) {
+    /**
+     * Check if the provided node is an instance of the provided class. If yes, then add the node to the provided list. The provided class must extend JexlNode.
+     * 
+     * @see org.apache.commons.jexl2.parser.JexlNode;
```
here

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java:
```diff
@@ -1195,43 +1256,116 @@ public static LiteralRange<BigDecimal> getBigDecimalBoundedRange(JexlNode[] chil
+    /**
+     * Iterate through provided node's ancestors and return true if any are an instance of the provided class. The provided class must extend JexlNode.
+     * 
+     * @see org.apache.commons.jexl2.parser.JexlNode;
```
here

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java:
```diff
@@ -1195,43 +1256,116 @@ public static LiteralRange<BigDecimal> getBigDecimalBoundedRange(JexlNode[] chil
+     * Iterate through all descendants of the provided node and return true if any are an instance of the provided class. The provided class must extend
+     * JexlNode.
+     * 
+     * @see org.apache.commons.jexl2.parser.JexlNode;
```
here

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java:
your unit tests are failing due to some of your javadoc.  apparently javadoc doesn't like a `@see` tag that ends with a semicolon.  I pointed out the offenders below.

jack-gitdev replied:
Thanks. I removed them and it should build successfully now.

jack-gitdev committed:
Removed semicolons from Javadocs.

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java:
```diff
@@ -1195,43 +1256,116 @@ public static LiteralRange<BigDecimal> getBigDecimalBoundedRange(JexlNode[] chil
+     * @return True if any ancestor is an instance of the provided class. If not, then False.
+     */
+    public static <T extends JexlNode> boolean isDescendantOfNodeType(JexlNode node, Class<T> typeKey) {
+        if (null != node && null != node.jjtGetParent()) {
```
Instead of the recursive calls here, you could have a more memory efficient setup like this. I don't think you should make this change in this merge request, but maybe that's work doing as a separate unit of work. 
```
while (node.getParent() != null){
  node = node.getParent()
  if (node.isInstance(type)) {
    return true
  }
}
```

apmoriarty replied:
Created https://github.com/NationalSecurityAgency/datawave/issues/1946 to address this

apmoriarty committed:
Merge branch 'integration' into task/document-query-parsing

apmoriarty committed:
Merge branch 'integration' into task/document-query-parsing
